ports/esp32s2/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,199 @@
+            //prioritize matched frequencies so we don't needlessly take slots
+            timer_index = i;
+            break;
+        } else if (reserved_timer_freq[i] == 0) {
+            timer_index = i;
+            break;
+        }
+    }
+    if (timer_index == INDEX_EMPTY) {
+        // Running out of timers isn't pin related on ESP32S2 so we can't re-use error messages
+        mp_raise_ValueError(translate("No more timers available"));
+    }
+
+    // Find a viable channel
+    for (size_t i = 0; i < LEDC_CHANNEL_MAX; i++) {
+        if (reserved_channels[i] == INDEX_EMPTY) {
+            channel_index = i;
+            break;
+        }
+    }
+    if (channel_index == INDEX_EMPTY) {
+        mp_raise_ValueError(translate("No more channels available"));
+    }
+
+    // Run configuration
+    self->tim_handle.timer_num = timer_index;
+    self->tim_handle.duty_resolution = duty_bits;
+    self->tim_handle.freq_hz = frequency;
+    self->tim_handle.speed_mode = LEDC_LOW_SPEED_MODE;
+    self->tim_handle.clk_cfg = LEDC_AUTO_CLK;
+
+    if (ledc_timer_config(&(self->tim_handle)) != ESP_OK) {
+        mp_raise_ValueError(translate("Could not initialize timer"));
+    }
+
+    self->chan_handle.channel = channel_index;
+    self->chan_handle.duty = duty >> (16 - duty_bits);
+    self->chan_handle.gpio_num = pin->number;
+    self->chan_handle.speed_mode = LEDC_LOW_SPEED_MODE; // Only LS is allowed on ESP32-S2
+    self->chan_handle.hpoint = 0;
+    self->chan_handle.timer_sel = timer_index;
+
+    if (ledc_channel_config(&(self->chan_handle))) {
+        mp_raise_ValueError(translate("Could not initialize channel"));
+    }
+
+    // Make reservations
+    reserved_timer_freq[timer_index] = frequency;
+    reserved_channels[channel_index] = timer_index;
+
+    self->variable_frequency = variable_frequency;
+    self->pin_number = pin->number;
+    self->deinited = false;
+    self->duty_resolution = duty_bits;
+    claim_pin(pin);
+
+    // Set initial duty
+    ledc_set_duty(LEDC_LOW_SPEED_MODE, self->chan_handle.channel, duty >> (16 - duty_bits));
+    ledc_update_duty(LEDC_LOW_SPEED_MODE, self->chan_handle.channel);
+
+    return PWMOUT_OK;
+}
+
+void common_hal_pulseio_pwmout_never_reset(pulseio_pwmout_obj_t *self) {
+    never_reset_tim[self->tim_handle.timer_num] = true;
+    never_reset_chan[self->chan_handle.channel] = true;
+}
+
+void common_hal_pulseio_pwmout_reset_ok(pulseio_pwmout_obj_t *self) {
+    never_reset_tim[self->tim_handle.timer_num] = false;
+    never_reset_chan[self->chan_handle.channel] = false;
+}
+
+bool common_hal_pulseio_pwmout_deinited(pulseio_pwmout_obj_t* self) {
+    return self->deinited == true;
+}
+
+void common_hal_pulseio_pwmout_deinit(pulseio_pwmout_obj_t* self) {
+    if (common_hal_pulseio_pwmout_deinited(self)) {
+        return;
+    }
+    ledc_stop(LEDC_LOW_SPEED_MODE, self->chan_handle.channel, 0);
+    ledc_timer_rst(LEDC_LOW_SPEED_MODE, self->tim_handle.timer_num);
```

Comment:
I don't think you want to do this if the timer is shared. Only later if `taken` is `false`.